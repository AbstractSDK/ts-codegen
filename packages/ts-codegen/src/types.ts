import { TSClientOptions, ReactQueryOptions } from "wasm-ast-types";

export interface TSBuilderOptions {
    contractDirs: string[];
    outPath: string;

    // TODO: separate types
    typesOnly?: boolean;

    reactQuery?: ReactQueryOptions & { enabled: true };
    tsClient?: TSClientOptions & { enabled: true };
};