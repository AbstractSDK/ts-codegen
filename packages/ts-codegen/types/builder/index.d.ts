import { TSClientOptions, ReactQueryOptions } from "wasm-ast-types";
export interface TSBuilderInput {
    contracts: Array<ContractFile | string>;
    outPath: string;
    options?: TSBuilderOptions;
}
export interface TSBuilderOptions {
    tsClient?: TSClientOptions & {
        enabled: boolean;
    };
    reactQuery?: ReactQueryOptions & {
        enabled: boolean;
    };
    recoil?: {
        enabled: boolean;
    };
    messageComposer?: {
        enabled: boolean;
    };
    bundle?: boolean;
}
export interface BuilderFile {
    contract: string;
    localname: string;
    filename: string;
}
export interface ContractFile {
    name: string;
    dir: string;
}
export declare class TSBuilder {
    contracts: Array<ContractFile | string>;
    outPath: string;
    options?: TSBuilderOptions;
    typesOnly?: boolean;
    protected tsClientFiles: BuilderFile[];
    protected recoilFiles: BuilderFile[];
    protected reactQueryFiles: BuilderFile[];
    protected messageComposerFiles: BuilderFile[];
    constructor({ contracts, outPath, options }: TSBuilderInput);
    getContracts(): ContractFile[];
    renderTsClient(contract: ContractFile): Promise<void>;
    renderRecoil(contract: ContractFile): Promise<void>;
    renderReactQuery(contract: ContractFile): Promise<void>;
    renderMessageComposer(contract: ContractFile): Promise<void>;
    build(): Promise<void>;
}
