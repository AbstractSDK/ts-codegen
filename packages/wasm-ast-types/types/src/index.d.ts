export * from './utils';
export * from './client';
export * from './recoil';
export * from './from-partial';
export * from './react-query';
