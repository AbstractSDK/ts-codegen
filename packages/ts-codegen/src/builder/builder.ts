import { TSClientOptions, ReactQueryOptions, defaultOptions } from "wasm-ast-types";
import { header } from '../utils/header';
import { join } from "path";
import { writeFileSync } from 'fs';
import { sync as mkdirp } from "mkdirp";
import messageComposer from '../generators/message-composer';
import reactQuery from '../generators/react-query';
import recoil from '../generators/recoil';
import tsClient from '../generators/ts-client';

import { basename } from 'path';
import { readSchemas } from '../utils';

import deepmerge from 'deepmerge';
import { pascal } from "case";
import { createFileBundle, recursiveModuleBundle } from "../bundler";

import generate from '@babel/generator';
import * as t from '@babel/types';

const defaultOpts: TSBuilderOptions = {
    tsClient: {
        enabled: true
    },
    recoil: {
        enabled: false
    },
    reactQuery: {
        enabled: false
    },
    messageComposer: {
        enabled: false
    },
    bundle: true
}

export interface TSBuilderInput {
    contracts: Array<ContractFile | string>;
    outPath: string;
    options?: TSBuilderOptions;
};

export interface TSBuilderOptions {
    tsClient?: TSClientOptions & { enabled: boolean };
    reactQuery?: ReactQueryOptions & { enabled: boolean };
    recoil?: { enabled: boolean };
    messageComposer?: { enabled: boolean };
    ///// END PLUGINS

    bundle?: boolean;
};

export interface BuilderFile {
    contract: string;
    localname: string;
    filename: string;
};

export interface ContractFile {
    name: string;
    dir: string;
}
export class TSBuilder {
    contracts: Array<ContractFile | string>;
    outPath: string;
    options?: TSBuilderOptions;

    // TODO: separate types
    typesOnly?: boolean;

    protected tsClientFiles: BuilderFile[] = [];
    protected recoilFiles: BuilderFile[] = [];
    protected reactQueryFiles: BuilderFile[] = [];
    protected messageComposerFiles: BuilderFile[] = [];

    constructor({ contracts, outPath, options }: TSBuilderInput) {
        this.contracts = contracts;
        this.outPath = outPath;
        this.options = deepmerge(
            deepmerge(
                defaultOptions,
                defaultOpts
            ),
            options ?? {}
        );
    }

    getContracts(): ContractFile[] {
        return this.contracts.map(contractOpt => {
            if (typeof contractOpt === 'string') {
                const name = basename(contractOpt);
                const contractName = pascal(name);
                return {
                    name: contractName,
                    dir: contractOpt
                }
            }
            return {
                name: pascal(contractOpt.name),
                dir: contractOpt.dir
            };
        });
    }

    async renderTsClient(contract: ContractFile) {
        const { enabled, ...options } = this.options.tsClient;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        const files = await tsClient(contract.name, schemas, this.outPath, options);
        [].push.apply(this.tsClientFiles, files);
    }

    async renderRecoil(contract: ContractFile) {
        const { enabled, ...options } = this.options.recoil;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        const files = await recoil(contract.name, schemas, this.outPath, options);
        [].push.apply(this.recoilFiles, files);
    }

    async renderReactQuery(contract: ContractFile) {
        const { enabled, ...options } = this.options.reactQuery;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        const files = await reactQuery(contract.name, schemas, this.outPath, options);
        [].push.apply(this.reactQueryFiles, files);
    }

    async renderMessageComposer(contract: ContractFile) {
        const { enabled, ...options } = this.options.messageComposer;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        const files = await messageComposer(contract.name, schemas, this.outPath, options);
        [].push.apply(this.messageComposerFiles, files);
    }

    async build() {
        const contracts = this.getContracts();
        for (let c = 0; c < contracts.length; c++) {
            const contract = contracts[c];
            await this.renderTsClient(contract);
            await this.renderMessageComposer(contract);
            await this.renderReactQuery(contract);
            await this.renderRecoil(contract);
        }
    }

    async bundle() {

        const allFiles = [
            ...this.reactQueryFiles,
            ...this.recoilFiles,
            ...this.tsClientFiles,
            ...this.messageComposerFiles
        ];

        const bundleFile = 'bundle.ts';
        const bundleVariables = {};
        const importPaths = [];

        allFiles.forEach(file => {
            createFileBundle(
                `contracts.${file.contract}`,
                file.localname,
                bundleFile,
                importPaths,
                bundleVariables
            );

        });

        const ast = recursiveModuleBundle(bundleVariables);

        const code = generate(t.program(
            [
                ...importPaths,
                ...ast
            ]
        )).code;

        mkdirp(this.outPath);
        writeFileSync(join(this.outPath, bundleFile), code);

    }
}