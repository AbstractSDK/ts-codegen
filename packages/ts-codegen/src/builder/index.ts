import { TSClientOptions, ReactQueryOptions, defaultOptions } from "wasm-ast-types";

import fromPartial from '../generators/from-partial';
import reactQuery from '../generators/react-query';
import recoil from '../generators/recoil';
import tsClient from '../generators/ts-client';

import { basename } from 'path';
import { readSchemas } from '../utils';

import deepmerge from 'deepmerge';
import { pascal } from "case";

const defaultOpts = {
    tsClient: {
        enabled: true
    },
    recoil: {
        enabled: false
    },
    reactQuery: {
        enabled: false
    },
    messageComposer: {
        enabled: false
    }
}

export interface TSBuilderInput {
    contracts: Array<ContractFile | string>;
    outPath: string;
    options?: TSBuilderOptions;
};

export interface TSBuilderOptions {
    tsClient?: TSClientOptions & { enabled: true };
    reactQuery?: ReactQueryOptions & { enabled: true };
    recoil?: { enabled: true };
    messageComposer?: { enabled: true };
};

export interface BuilderFile {
    contract: string;
    localname: string;
    filename: string;
};

export interface ContractFile {
    name: string;
    dir: string;
}
export class TSBuilder {
    contracts: Array<ContractFile | string>;
    outPath: string;
    options?: TSBuilderOptions;

    // TODO: separate types
    typesOnly?: boolean;

    readonly typeFiles: BuilderFile[] = [];
    readonly tsClientFiles: BuilderFile[] = [];
    readonly recoilFiles: BuilderFile[] = [];
    readonly reactQueryFiles: BuilderFile[] = [];

    constructor({ contracts, outPath, options }: TSBuilderInput) {
        this.contracts = contracts;
        this.outPath = outPath;
        this.options = deepmerge(
            deepmerge(
                defaultOptions,
                defaultOpts
            ),
            options ?? {}
        );
    }

    getContracts(): ContractFile[] {
        return this.contracts.map(contractOpt => {
            if (typeof contractOpt === 'string') {
                const name = basename(contractOpt);
                const contractName = pascal(name);
                return {
                    name: contractName,
                    dir: contractOpt
                }
            }
            return {
                name: pascal(contractOpt.name),
                dir: contractOpt.dir
            };
        });
    }

    async renderTsClient(contract: ContractFile) {
        const { enabled, ...options } = this.options.tsClient;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        await tsClient(contract.name, schemas, this.outPath, options);
    }

    async renderRecoil(contract: ContractFile) {
        const { enabled, ...options } = this.options.recoil;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        await recoil(contract.name, schemas, this.outPath, options);
    }

    async renderReactQuery(contract: ContractFile) {
        const { enabled, ...options } = this.options.reactQuery;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        await reactQuery(contract.name, schemas, this.outPath, options);
    }

    async renderMessageComposer(contract: ContractFile) {
        const { enabled, ...options } = this.options.messageComposer;
        if (!enabled) return;
        const schemas = await readSchemas({ schemaDir: contract.dir });
        await fromPartial(contract.name, schemas, this.outPath, options);
    }

    async build() {
        const contracts = this.getContracts();
        for (let c = 0; c < contracts.length; c++) {
            const contract = contracts[c];
            await this.renderTsClient(contract);
            await this.renderMessageComposer(contract);
            await this.renderReactQuery(contract);
            await this.renderRecoil(contract);
        }
    }
}