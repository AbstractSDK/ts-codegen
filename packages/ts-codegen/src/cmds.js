
import _generate from './commands/generate';
import _message_composer from './commands/message-composer';
import _react_query from './commands/react-query';
import _recoil from './commands/recoil';
import _types from './commands/types';
const Commands = {};
Commands['generate'] = _generate;
Commands['message-composer'] = _message_composer;
Commands['react-query'] = _react_query;
Commands['recoil'] = _recoil;
Commands['types'] = _types;

  export { Commands }; 

  