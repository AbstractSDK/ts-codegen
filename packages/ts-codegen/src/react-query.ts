import { pascal } from "case";
import { header } from './header';
import { join } from "path";
import { sync as mkdirp } from "mkdirp";
import * as w from 'wasm-ast-types';
import { RenderContext } from 'wasm-ast-types';
import * as t from '@babel/types';
import { writeFileSync } from 'fs';
import generate from "@babel/generator";
import { findAndParseTypes, findExecuteMsg, findQueryMsg, getDefinitionSchema } from './utils';
import { getMessageProperties, ReactQueryOptions } from "wasm-ast-types";
import { cosmjsAminoImportStatements } from './imports';

export default async (
    contractName: string,
    schemas: any[],
    outPath: string,
    reactQueryOptions?: ReactQueryOptions
) => {
    const context = new RenderContext(getDefinitionSchema(schemas), {
        reactQuery: reactQueryOptions ?? {}
    });
    const options = context.options.reactQuery;

    const ReactQueryFile = pascal(`${contractName}Contract`) + '.react-query.ts';
    const Contract = pascal(`${contractName}Contract`)

    const QueryMsg = findQueryMsg(schemas);
    const ExecuteMsg = findExecuteMsg(schemas);
    const typeHash = await findAndParseTypes(schemas);

    const ExecuteClient = pascal(`${contractName}Client`);
    const QueryClient = pascal(`${contractName}QueryClient`);

    const body = [];

    const reactQueryImports = ['useQuery', 'UseQueryOptions']
    const clientImports = []

    QueryMsg && clientImports.push(QueryClient)

    // check that there are commands within the exec msg
    const shouldGenerateMutationHooks = ExecuteMsg && options?.v4 && options?.mutations && getMessageProperties(ExecuteMsg).length > 0

    if (shouldGenerateMutationHooks) {
        body.push(w.importStmt(['ExecuteResult'], '@cosmjs/cosmwasm-stargate'));
        body.push(cosmjsAminoImportStatements(typeHash))
        reactQueryImports.push('useMutation', 'UseMutationOptions')
        clientImports.push(ExecuteClient)
    }

    // react-query imports
    body.push(
        w.importStmt(reactQueryImports, options?.v4 ? '@tanstack/react-query' : 'react-query')
    );

    // general contract imports
    body.push(w.importStmt(Object.keys(typeHash), `./${Contract}`));

    // client imports
    body.push(w.importStmt(clientImports, `./${Contract}`));

    // query messages
    if (QueryMsg) {
        [].push.apply(body,
            w.createReactQueryHooks({
                context,
                queryMsg: QueryMsg,
                contractName: contractName,
                QueryClient
            })
        );
    }

    if (shouldGenerateMutationHooks) {
        [].push.apply(body,
            w.createReactQueryMutationHooks({
                context,
                execMsg: ExecuteMsg,
                contractName: contractName,
                ExecuteClient
            })
        );
    }



    const code = header + generate(
        t.program(body)
    ).code;

    mkdirp(outPath);
    writeFileSync(join(outPath, ReactQueryFile), code);
};
