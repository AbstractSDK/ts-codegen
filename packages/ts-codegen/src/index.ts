import { TSBuilder, TSBuilderInput } from './builder';

export { default as tsClient } from './generators/ts-client';
export { default as messageComposer } from './generators/message-composer';
export { default as reactQuery } from './generators/react-query';
export { default as recoil } from './generators/recoil';

export * from './utils';
export * from './utils/imports';
export * from './builder';
export * from './bundler';

export default async (input: TSBuilderInput) => {
    const builder = new TSBuilder(input);
    await builder.build()
};

