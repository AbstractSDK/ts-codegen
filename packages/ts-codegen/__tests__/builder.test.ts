import { TSBuilder } from '../src/builder';

const FIXTURE_DIR = __dirname + '/../../../__fixtures__';
const OUTPUT_DIR = __dirname + '/../../../__output__/builder';

it('options undefined', async () => {
    const outPath = OUTPUT_DIR + '/vectis/factory-opt';
    const schemaDir = FIXTURE_DIR + '/vectis/factory/';
    const builder = new TSBuilder({
        contracts: [
            schemaDir
        ],
        outPath,
        options: undefined
    });
    expect(builder).toMatchSnapshot();
});

it('options tsClient.enabled', async () => {
    const outPath = OUTPUT_DIR + '/vectis/factory-opt';
    const schemaDir = FIXTURE_DIR + '/vectis/factory/';
    const builder = new TSBuilder({
        contracts: [
            schemaDir
        ],
        outPath,
        options: {
            tsClient: {
                enabled: true
            },
            reactQuery: {
                enabled: true
            }
        }
    });
    expect(builder).toMatchSnapshot();
});

it('builder invoke', async () => {
    const outPath = OUTPUT_DIR + '/invoke/';
    const s = (str) => FIXTURE_DIR + str;
    const builder = new TSBuilder({
        contracts: [
            s('/vectis/factory'),
            s('/minter'),
            s('/daodao/cw-admin-factory'),
            s('/daodao/cw-code-id-registry'),
            {
                name: 'CwSingle',
                dir: s('/daodao/cw-proposal-single')
            }
        ],
        outPath,
        options: {
            tsClient: {
                enabled: true
            },
            reactQuery: {
                enabled: true
            }
        }
    });
    await builder.build();
    await builder.bundle();
});
