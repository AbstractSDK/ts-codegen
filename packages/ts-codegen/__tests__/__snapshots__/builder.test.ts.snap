// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`options tsClient.enabled 1`] = `
TSBuilder {
  "files": Array [],
  "options": Object {
    "bundle": Object {
      "bundleFile": "bundle.ts",
      "enabled": true,
      "scope": "contracts",
    },
    "client": Object {
      "enabled": true,
    },
    "messageComposer": Object {
      "enabled": false,
    },
    "reactQuery": Object {
      "camelize": true,
      "enabled": true,
      "mutations": false,
      "optionalClient": false,
      "version": "v3",
    },
    "recoil": Object {
      "enabled": false,
    },
    "types": Object {
      "aliasExecuteMsg": false,
      "enabled": true,
    },
  },
  "outPath": "/Users/pyramation/code/cosmwasm/ts-codegen/packages/ts-codegen/__tests__/../../../__output__/builder/vectis/factory-opt",
}
`;

exports[`options undefined 1`] = `
TSBuilder {
  "files": Array [],
  "options": Object {
    "bundle": Object {
      "bundleFile": "bundle.ts",
      "enabled": true,
      "scope": "contracts",
    },
    "client": Object {
      "enabled": true,
    },
    "messageComposer": Object {
      "enabled": false,
    },
    "reactQuery": Object {
      "camelize": true,
      "enabled": false,
      "mutations": false,
      "optionalClient": false,
      "version": "v3",
    },
    "recoil": Object {
      "enabled": false,
    },
    "types": Object {
      "aliasExecuteMsg": false,
      "enabled": true,
    },
  },
  "outPath": "/Users/pyramation/code/cosmwasm/ts-codegen/packages/ts-codegen/__tests__/../../../__output__/builder/vectis/factory-opt",
}
`;
